packages/asset-swapper/src/utils/quote_report_generator.ts

```diff
@@ -116,6 +137,85 @@ export function generateQuoteReport(
     };
 }
 
+/**
+ * Generates a report of sources considered while computing the optimized swap quote, the sources ultimately included in the computed quote, and all considered quotes.
+ */
+export function generateExtendedQuoteReport(
+    marketOperation: MarketOperation,
+    quotes: RawQuotes,
+    liquidityDelivered: ReadonlyArray<CollapsedFill> | DexSample<MultiHopFillData>,
+    amount: BigNumber,
+    comparisonPrice?: BigNumber | undefined,
+    quoteRequestor?: QuoteRequestor,
+): ExtendedQuoteReport {
+    const sourcesConsidered: ExtendedQuoteReportEntry[] = [];
+
+    // NativeOrders
+    sourcesConsidered.push(
+        ...quotes.nativeOrders.map(order =>
+            nativeOrderToReportEntry(
+                order.type,
+                order as any,
+                order.fillableTakerAmount,
+                comparisonPrice,
+                quoteRequestor,
+            ),
+        ),
+    );
+
+    // IndicativeQuotes
+    sourcesConsidered.push(
+        ...quotes.rfqtIndicativeQuotes.map(order => indicativeQuoteToReportEntry(order, comparisonPrice)),
+    );
+
+    // MultiHop
+    sourcesConsidered.push(...quotes.twoHopQuotes.map(quote => multiHopSampleToReportSource(quote, marketOperation)));
+
+    // Dex Quotes
+    sourcesConsidered.push(
+        ..._.flatten(
+            quotes.dexQuotes.map(dex =>
+                dex
+                    .filter(quote => isDexSampleForTotalAmount(quote, marketOperation, amount))
+                    .map(quote => dexSampleToReportSource(quote, marketOperation)),
+            ),
+        ),
```

Comment:
Can this be simplified using `flatMap`? 
```
        ...quotes.dexQuotes.flatMap(dex =>
            dex
                .filter(quote => isDexSampleForTotalAmount(quote, marketOperation, amount))
                .map(quote => dexSampleToReportSource(quote, marketOperation)),
        ),

Reply:
I didn't know about `flatMap` sounds much better, but currently the whole `protocol` repo uses `es2017` and flatMap was introduced in `es2019`. I would prefer not to change it in this PR.